src/modules/git_commit.rs

```diff
@@ -98,6 +98,40 @@ mod tests {
         repo_dir.close()
     }
 
+    #[test]
+    fn test_ceiling_directory() -> io::Result<()> {
+        let repo_dir = fixture_repo(FixtureProvider::Git)?;
+        let level1 = repo_dir.path().join("level1");
+        let level2 = level1.join("level2");
+        std::fs::create_dir(&level1)?;
+        std::fs::create_dir(&level2)?;
+
+        const CEILING_DIRECTORY_VAR: &str = "GIT_CEILING_DIRECTORIES";
+
+        let prev_ceiling = std::env::var_os(CEILING_DIRECTORY_VAR);
+        #[allow(clippy::disallowed_methods)]
+        std::env::set_var(CEILING_DIRECTORY_VAR, level1.as_os_str());
```

Comment:
I think it might better to append the env value instead of replacing and restoring it, maybe even adding it twice to test your upstream PR (thanks for fixing the upstream bug I introduced!).
Beyond that, shouldn't the tests already pass on Windows? You are only adding a single path (so there is no platform-specific delimiter), there might be more issues here.

Reply:
> Beyond that, shouldn't the tests already pass on Windows?

No, since the paths on Windows include the Drive letter (e.g. `C:\Users\User` → `["C", "\Users\User"]`).

> You are only adding a single path (so there is no platform-specific delimiter), there might be more issues here.

I could add more test-cases. If the tests are running concurrently, this could introduce flaky tests.

> I think it might better to append the env value instead of replacing and restoring it, maybe even adding it twice to test your upstream PR (thanks for fixing the upstream bug I introduced!).

Wouldn't that make the tests dependent on the outside environment?

Reply:
Ok, that sounds fine.

> Wouldn't that make the tests dependent on the outside environment?

They were already included by the environment to begin with, and only changing the environment once feels less disruptive and makes any potential impact more observable.

Reply:
Hmm, I can't really get the appending to work.

<details>
<summary>My current approach</summary>

```diff
diff --git a/src/modules/git_commit.rs b/src/modules/git_commit.rs
index f9a18bc2..ada24996 100644
--- a/src/modules/git_commit.rs
+++ b/src/modules/git_commit.rs
@@ -98,6 +98,22 @@ mod tests {
         repo_dir.close()
     }
 
+    #[allow(clippy::disallowed_methods)]
+    fn append_ceiling_directory(directory: &std::ffi::OsStr) {
+        const CEILING_DIRECTORY_VAR: &str = "GIT_CEILING_DIRECTORIES";
+
+        let ceiling: std::borrow::Cow<std::ffi::OsStr> =
+            match std::env::var_os(CEILING_DIRECTORY_VAR) {
+                Some(mut var) => {
+                    var.push(if cfg!(windows) { ";" } else { ":" });
+                    var.push(directory);
+                    var.into()
+                }
+                None => directory.into(),
+            };
+        std::env::set_var(CEILING_DIRECTORY_VAR, ceiling);
+    }
+
     #[test]
     fn test_ceiling_directory() -> io::Result<()> {
         let repo_dir = fixture_repo(FixtureProvider::Git)?;
@@ -106,11 +122,8 @@ mod tests {
         std::fs::create_dir(&level1)?;
         std::fs::create_dir(&level2)?;
 
-        const CEILING_DIRECTORY_VAR: &str = "GIT_CEILING_DIRECTORIES";
-
-        let prev_ceiling = std::env::var_os(CEILING_DIRECTORY_VAR);
-        #[allow(clippy::disallowed_methods)]
-        std::env::set_var(CEILING_DIRECTORY_VAR, level1.as_os_str());
+        append_ceiling_directory(level1.as_os_str());
+        append_ceiling_directory(level1.as_os_str());
 
         let actual = ModuleRenderer::new("git_commit")
             .config(toml::toml! {
@@ -122,11 +135,40 @@ mod tests {
 
         let expected = None;
 
-        match prev_ceiling {
-            #[allow(clippy::disallowed_methods)]
-            Some(ceiling) => std::env::set_var(CEILING_DIRECTORY_VAR, ceiling),
-            None => std::env::remove_var(CEILING_DIRECTORY_VAR),
-        };
+        assert_eq!(expected, actual);
+        repo_dir.close()
+    }
+
+    #[test]
+    fn test_ceiling_directory_in_current_dir() -> io::Result<()> {
+        let repo_dir = fixture_repo(FixtureProvider::Git)?;
+        let level1 = repo_dir.path().join("level1");
+        let level2 = level1.join("level2");
+        std::fs::create_dir(&level1)?;
+        std::fs::create_dir(&level2)?;
+
+        append_ceiling_directory(repo_dir.path().as_os_str());
+
+        let mut git_output = create_command("git")?
+            .args(["rev-parse", "HEAD"])
+            .current_dir(repo_dir.path())
+            .output()?
+            .stdout;
+        git_output.truncate(7);
+        let expected_hash = str::from_utf8(&git_output).unwrap();
+
+        let actual = ModuleRenderer::new("git_commit")
+            .config(toml::toml! {
+                [git_commit]
+                    only_detached = false
+            })
+            .path(repo_dir.path())
+            .collect();
+
+        let expected = Some(format!(
+            "{} ",
+            Color::Green.bold().paint(format!("({expected_hash})"))
+        ));
 
         assert_eq!(expected, actual);
         repo_dir.close()
```
</details>

After setting the ceiling directory to `C:\Users\User\AppData\Local\Temp\.tmpaSiHpn`, no other git repository can be discovered - even if the tempdir has a different name (e.g. `C:\Users\Johannes\AppData\Local\Temp\.tmpDHMU0f`), when running `cargo test --lib -- modules::git_commit::tests::test_render_commit_hash modules::git_commit::tests::test_ceiling_directory_in_current_dir --exact --nocapture `

Reply:
I think that's a different issue (Byron/gitoxide#721). Still, I would prefer only setting the variable once to avoid any possible race-conditions between reading and writing the environment variable.

Reply:
You mean there should be a single test that sets the variable and tests multiple things with it?

Reply:
Yes.